### ISSUE TITLE
generic class attributes with unions become object on subclasses

### ISSUE THREAD

**Issue by chadrik** [python/mypy#4834]
I'm not sure if this is a bug or not, but if it is by design it would be great if it was covered in the docs, as I found it confusing:

```python
from typing import *

class A:
    a = None  # type: List[str]
    b = None  # type: List[Union[str, int]]
    c = None  # type: List[Union[str, int]]
    d = None  # type: Sequence[Union[str, int]]

class B(A):
    a = ['foo']
    b = ['foo']
    c = ['foo', 1]
    d = ['foo', 1]

reveal_type(B.a)  # 'builtins.list[builtins.str*]'
reveal_type(B.b)  # 'builtins.list[builtins.str*]'
reveal_type(B.c)  # 'builtins.list[builtins.object*]'
reveal_type(B.d)  # 'builtins.list[builtins.object*]'
```

I expected `c` and `d` to remain unions.

**Comment by ilevkivskyi** [comment]
What I see here is a bigger bug, namely it looks like we are currently don't check Liskov for inferred types. I think this may be because we currently don't have backtracking in our inference engine, so currently we will have some false positives. Another thing that we can do is to use explicit (i.e. declared) or even implicit (i.e. inferred) suprtype type as a context for inference. This however may lead to wider types than one would expect. So that I think ultimately we may need both backtracking and using supertype context to conveniently (for user) treat such situations.

These are however two very different scale tasks, so I would propose to just go with using explicit supertype context (this seems to be pretty easy). Actually, if we are going to prohibit covariant subtyping of mutable attributes at some point, just using supertype context would be the final/full solution. @JukkaL what do you think?

**Comment by chadrik** [comment]
> Another thing that we can do is to use explicit (i.e. declared) or even implicit (i.e. inferred) suprtype type as a context for inference. This however may lead to wider types than one would expect. 

Personally I think it's desirable for the types of attributes on the subclasses to remain the same as their parents unless they are explicitly redeclared.  e.g. Even in the case of `B.b` above  I would expect the type to be `List[Union[str, int]]`.

A related issue that I've found inconvenient is that variable length tuples (e.g. `Tuple[str, ...]`) on a parent class attribute will become an explicit length (e.g. `Tuple[str, str]`) on subtypes.  This effectively means that I need to specify the type for all child attributes.

**Comment by JukkaL** [comment]
> What I see here is a bigger bug, namely it looks like we are currently don't check Liskov for inferred types.

Yes, this is a known issue -- but I can't an existing GitHub issue for this. I have a vague recollection of looking at this and I suspect fixing this could trigger a bunch of errors in Dropbox internal codebases. Having support for final/constant attributes would help with this, as they could vary covariantly.

> These are however two very different scale tasks, so I would propose to just go with using explicit supertype context (this seems to be pretty easy). Actually, if we are going to prohibit covariant subtyping of mutable attributes at some point, just using supertype context would be the final/full solution. @JukkaL what do you think?

I think that we should use the type of base class attribute as the context, no matter if it's explicitly declared or inferred. Declared and inferred types are treated the same elsewhere so for consistency we should do the same here.

**Comment by ilevkivskyi** [comment]
This is now fixed (we use supertype as inference context).
